HIP/hip-376.md

```diff
@@ -0,0 +1,156 @@
+
+The following `ERC-20` operations will be supported. Standard `ERC-20 Events` will be emitted as appropriate.
+
+- `function allowance(address owner, address spender) external view returns (uint256);`
+- `function approve(address spender, uint256 amount) external returns (bool);`
+- `function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);`
+
+#### Supported ERC-721 operations
+
+Tokens of type `NON_FUNGIBLE_UNIQUE` will support standard `ERC-721` calls.
+
+The following `ERC-721` operations will be supported. Standard `ERC-721 Events` will be emitted as appropriate.
+
+  - `function approve(address _approved, uint256 _tokenId) external payable`
+  - `function setApprovalForAll(address _operator, bool _approved) external`
+  - `function getApproved(uint256 _tokenId) external view returns (address)`
+  - `function isApprovedForAll(address _owner, address _operator) external view returns (bool)`
+  - `function transferFrom(address _from, address _to, uint256 _tokenId) external payable`
+
+#### ERC20 and ERC721 mapping to the HAPI interface
+
+There are several differences in the way `ERC20` and `ERC721` functions are designed to behave and the way HAPI transactions for `approve`, `allowance` and `transferFrom` are implemented in Hedera. This section describes the integration between the function specifications and the actual Hedera behaviour.
+
+##### ERC20 approve and HAPI Adjust allowance
+
+ERC20 `approve` function works by overriding the previous value (if any was set) for the 'spender' address specified.
+HAPI `CryptoAdjustAllowance` transactions work by applying the difference specified by the `amount` field in the transaction body to the existing allowance (or 0 if there is no existing allowance approved for that account).
+Calling ERC20 `approve` with amount `200` will be implemented in the following way in the precompile:
+ - query the current configured allowance for the `owner`, `spender` and `tokenId`
+ - calculate the adjustment that needs to be made to have the final allowance equal to the amount received as argument to the ERC20 `approve`
+ - create a `CryptoAdjustAllowance` tx with the calculated adjustment
+
+##### ERC721 approve and multiple approved accounts for NFT in Hedera
+
+ERC721 is designed in such a way that an NFT can have either 0 or 1 approvals e.g. if Alice owns NFT 'X' and Alice calls the approve function two times, once with Bob's account and once with Carol's account, the second call to approve will override Bob's allowance and only Carol will have approved allowance for that NFT. In Hedera the behaviour is different and both Bob and Carol would have approved allowances after the second call to approve.
+To make sure there is no unexpected result for smart contract developers working with HTS tokens, ERC721 `approve` and `getApproved` will behave in the following way with HTS tokens:
+When `approve` is called:
```

Comment:
I have 3 concerns related to the proposed API behaviour.
1. We have potentially unbounded computation or most probably limited to the number of maximum `granted nft allowances`. This should be accounted in the `gas` costs of the precompile
2. It is common for users to change the approvals for specific tokens sequentially, meaning:
```
0. User A is the owner of tokenId=x
1. User A sets approval to User B
2. User A sets approval to User C
```

as per the specification, we will revert on step 2.

3. In the ERC721 standard, allowances are cleared on every transfer. It might be a surprise for developers to see that approvals are not cleared on transfers.

Reply:
1. Answered in the other comment reply
2. I think we can update this step from the specification:
`if no, then we'll check if the owner account has exactly 0 existing approvals for that NFT serial number`
to:
`if no, then we'll check if the owner account has exactly 0 or 1 existing approvals for that NFT serial number`
and for the case there is 1 existing approval granted, to replace it (revoke old + grant new)
3. Just checked behaviour in Hedera and allowances are not cleared on transfer, even though the account with the granted allowance is no longer the owner of the token. We could add logic for revoking grants on `transferFrom`. Main concerns are this would result in higher gas price and different behaviour than HAPI `cryptoTransfer` flow.